### PR TITLE
Some Dynamic Pal Improvements

### PR DESCRIPTION
**Pull request by vulcandth**
* No longer run Dynamic Pal in `LoadMapPals`
* Don't update `OBJECT_PALETTE` during the first dyn pal pass.
* Better Menu exit hook (Doesn't corrupt BG)
* Ensure the correct Wram bank is loaded.
* Preserve more registers, just in case.
* Re-fix Battle Transitions
* Re-fix sprites appearing before warp is complete
* Fix sprite problems when activating surf (Menu problem).
* Verified pals updating correctly when crossing map connections with different map conditions (Example: Overcast)

### CODE REVIEW

File: engine/gfx/dynamic_pals.asm [hunk 1]
```diff
@@ -1,20 +1,32 @@
...
 
 DisableDynPalUpdates::
+	push hl
 	ld hl, wPalFlags
 	set DISABLE_DYN_PAL_F, [hl]
+	pop hl
 	ret
 
 EnableDynPalUpdates::
+	push hl
```
Comment (Rangi42) [inline engine/gfx/dynamic_pals.asm:26]: `CheckForUsedObjPals` doesn't preserve `hl`, so either this is unnecessary or `CheckForUsedObjPals` should do it too.

File: home/map.asm [hunk 2]
```diff
@@ -1640,6 +1640,8 @@ ReturnToMapWithSpeechTextbox::
 	ld [wSpriteUpdatesEnabled], a
 	call ClearBGPalettes
 	call ClearSprites
+	farcall ClearSavedObjPals
```
Comment (Rangi42) [inline home/map.asm:1643]: After FIQ and ariscop were discussing `farcall`s, I wonder if this adds any perceptible delay to the overworld textboxes.

File: engine/gfx/dynamic_pals.asm (hunk 1 above)
Comment (Rangi42) [reply engine/gfx/dynamic_pals.asm:26]: `CheckForUsedObjPals` already does. <_<

File: home/map.asm (hunk 2 above)
Comment (ariscop) [reply home/map.asm:1643]: Perhaps the discussion in discord has overstated the problem; they're on the order of 10x the overhead of a plain call/ret, which isn't that bad in practice

ReloadTilesetAndPalettes already does a lot of farcalls, a few more is unlikely to be noticeable

**Pull request merged**